src/original_source.rs

```diff
@@ -0,0 +1,258 @@
+use std::{
+  borrow::Cow,
+  hash::{Hash, Hasher},
+};
+
+use crate::{
+  helpers::{
+    get_generated_source_info, get_map, split_into_lines,
+    split_into_potential_tokens, GeneratedInfo, OnChunk, OnName, OnSource,
+    StreamChunks,
+  },
+  source::{Mapping, OriginalLocation},
+  MapOptions, Source, SourceMap,
+};
+
+#[derive(Debug, Clone)]
+pub struct OriginalSource {
+  value: String,
+  name: String,
+}
+
+impl OriginalSource {
+  pub fn new(value: impl Into<String>, name: impl Into<String>) -> Self {
+    Self {
+      value: value.into(),
+      name: name.into(),
+    }
+  }
+}
+
+impl Source for OriginalSource {
+  fn source(&self) -> Cow<str> {
+    Cow::Borrowed(&self.value)
+  }
+
+  fn buffer(&self) -> Cow<[u8]> {
+    Cow::Borrowed(self.value.as_bytes())
+  }
+
+  fn size(&self) -> usize {
+    self.value.len()
+  }
+
+  fn map(&self, options: &MapOptions) -> Option<SourceMap> {
+    get_map(self, options)
+  }
+}
+
+impl Hash for OriginalSource {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    "OriginalSource".hash(state);
```

Comment:
`type_name_of_val` maybe a better solution?

Reply:
`type_name_of_val` seems not appropriate for this.

> The type name should not be considered a unique identifier of a type; multiple types may share the same type name.

according to https://doc.rust-lang.org/std/any/fn.type_name_of_val.html#note

Reply:
My main concern is that whether `type_name_of_val` or `type_name` will return different str on different os/arch, according to

> For example, amongst the strings that type_name::<Option<String>>() might return are "Option<String>" and "std::option::Option<std::string::String>".

And I worry it might cause potential problem which will be very difficult to find out what's causing the problem.

Reply:
LGTM for keeping it as-is.